docs/reference/antiquoted-c.rst

```diff
@@ -16,114 +19,173 @@ more capability is required.
-    openf ($ty:(U32) arg)
-    {
-        /* ... */
-    }
+At compilation,
+|cogent| will replace the ``$ty`` term with
+a corresponding C type,
+so the generated result would be::
 
-  This will be transformed into::
+  u32 u32_from_u8(u8 x) { return x; }
 
-    t1 openf(u32 arg) { /* ... */ }
 
-  where the type definition ``t1`` has already been generated.
-  Beware that the |cogent| compiler
-  may not always emit the type's definition,
-  even if it emits a reference to it.
+This works for less-trivial types as well.
+In |cogent|,
+if we declare the types ``Result``, ``FileDesc``, ``Errno``,
+and a function of them, ``openf``::
 
+  type Result t e = < Ok t | Err e >
+  type FileDesc   = U32
+  type Errno      = U32
+  openf: U32 -> Result FileDesc Errno
 
-- ``$exp:(EXPRESSION)`` inserts code to execute ``EXPRESSION``.
-  The most common case for this is to call |cogent| functions,
-  or to access top-level bindings.
+We can then implement this function in C::
 
-  For example,
-  if we have a |cogent| function::
+  $ty:(Result FileDesc Errno)
+  openf ($ty:(U32) arg)
+  {
+      /* ... */
+  }
 
-    within_bounds: U32 -> Bool
-    within_bounds x = {- ... -}
+The compiler will transform this into something like::
 
-  Then, if we wish to invoke it from in C,
-  we might say::
+  t1 openf(u32 arg) { /* ... */ }
 
-    if ($exp:(within_bounds)(value)) { /* ... */ }
+where the type definition ``t1`` was previously generated.
+Beware that the |cogent| compiler
+may not always emit the type's definition,
+even if it emits a reference to it.
 
 
-- ``$spec:(TYPE-NAME)`` will,
-  if ``TYPE-NAME`` is a function,
-  insert an appropriate C type;
-  this is intended for use
-  when type-casting function pointers,
-  as is necessary when invoking a lambda.
+``$exp``: Embedding Expressions
+------------------------------------
 
+``$exp:(EXPRESSION)`` inserts code to execute ``EXPRESSION``.
+The most common case for this is to call |cogent| functions,
+or to access top-level bindings.
 
-- ``$esc:(PREPROCESSOR-DIRECTIVE)`` and
-  ``$escstm:(PREPROCESSOR-DIRECTIVE)``
-  allow insertion of a literal C preprocessor directive.
+For example,
+if we have a |cogent| function::
+
+  within_bounds: U32 -> Bool
+  within_bounds x = {- ... -}
+
+Then, if we wish to invoke it from in C,
+we might say::
+
+  if ($exp:(within_bounds)(value)) { /* ... */ }
+
+
+``$spec``: Specifying Casts
+------------------------------------
+
+``$spec:(TYPE-NAME)`` will,
+if ``TYPE-NAME`` is the type of a function,
+insert an appropriate C typecast.
+Its most common use is
+when type-casting function pointers,
+as is necessary when invoking
+a function reference passed from |cogent| to C.
+
+As an example,
+let's consider a function whose type is
+
+
+  type ArrayModifyF a acc = OptElemA a acc -> OptElemA a acc
+
+
+
+``$esc`` and ``$escstm``: No Escape In Sight
+--------------------------------------------
+
+``$esc:(PREPROCESSOR-DIRECTIVE)`` and
```

Comment:
The thing inside the `$esc` and `$escstm` doesn't have to be a cpp directive. It can be anything that the user doesn't want the Cogent compiler to see. The only different is that `$esc` is top-level, and `$escstm` is on the statement level. The only criterion is that if you remove the entire `$esc` and $escstm` antiquotes, the rest should still be valid antiquote C.